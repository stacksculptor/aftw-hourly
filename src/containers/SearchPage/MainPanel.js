import React, { Component } from 'react';
import { array, bool, func, object, number, string } from 'prop-types';
import { FormattedMessage } from 'react-intl';
import classNames from 'classnames';
import { propTypes } from '../../util/types';
import {
  SearchResultsPanel,
  SearchFilters,
  SearchFiltersMobile,
  SearchFiltersPanel,
} from '../../components';

import css from './SearchPage.css';

class MainPanel extends Component {
  constructor(props) {
    super(props);
    this.state = { isSearchFiltersPanelOpen: false };
  }

  render() {
    const {
      className,
      rootClassName,
      urlQueryParams,
      listings,
      searchInProgress,
      searchListingsError,
      searchParamsAreInSync,
      onActivateListing,
      onManageDisableScrolling,
      onOpenModal,
      onCloseModal,
      onMapIconClick,
      pagination,
      searchParamsForPagination,
      showAsModalMaxWidth,
      customURLParamToConfig,
      primaryFilters,
      secondaryFilters,
    } = this.props;

    const isSearchFiltersPanelOpen = !!secondaryFilters && this.state.isSearchFiltersPanelOpen;
    const searchFiltersPanelSelectedCount = !secondaryFilters
      ? 0
      : Object.keys(customURLParamToConfig)
          .map(key => urlQueryParams[key])
          .filter(param => !!param).length;

    const searchFiltersPanelProps = !!secondaryFilters
      ? {
          isSearchFiltersPanelOpen: this.state.isSearchFiltersPanelOpen,
          toggleSearchFiltersPanel: isOpen => {
            this.setState({ isSearchFiltersPanelOpen: isOpen });
          },
          searchFiltersPanelSelectedCount,
        }
      : {};

    const hasPaginationInfo = !!pagination && pagination.totalItems != null;
    const totalItems = searchParamsAreInSync && hasPaginationInfo ? pagination.totalItems : 0;
    const listingsAreLoaded = !searchInProgress && searchParamsAreInSync && hasPaginationInfo;

    const classes = classNames(rootClassName || css.searchResultContainer, className);

    return (
      <div className={classes}>
        <SearchFilters
          className={css.searchFilters}
          urlQueryParams={urlQueryParams}
          listingsAreLoaded={listingsAreLoaded}
          resultsCount={totalItems}
          searchInProgress={searchInProgress}
          searchListingsError={searchListingsError}
          onManageDisableScrolling={onManageDisableScrolling}
          {...searchFiltersPanelProps}
          {...primaryFilters}
        />
        <SearchFiltersMobile
          className={css.searchFiltersMobile}
          urlQueryParams={urlQueryParams}
          listingsAreLoaded={listingsAreLoaded}
          resultsCount={totalItems}
          searchInProgress={searchInProgress}
          searchListingsError={searchListingsError}
          showAsModalMaxWidth={showAsModalMaxWidth}
          onMapIconClick={onMapIconClick}
          onManageDisableScrolling={onManageDisableScrolling}
          onOpenModal={onOpenModal}
          onCloseModal={onCloseModal}
          {...primaryFilters}
          {...secondaryFilters}
        />
        {isSearchFiltersPanelOpen ? (
          <div className={classNames(css.searchFiltersPanel)}>
            <SearchFiltersPanel
              urlQueryParams={urlQueryParams}
              customURLParamToConfig={customURLParamToConfig}
              listingsAreLoaded={listingsAreLoaded}
              onClosePanel={() => this.setState({ isSearchFiltersPanelOpen: false })}
              {...secondaryFilters}
            />
          </div>
        ) : (
          <div
            className={classNames(css.listings, {
              [css.newSearchInProgress]: !listingsAreLoaded,
            })}
          >
            {searchListingsError ? (
              <h2 className={css.error}>
                <FormattedMessage id="SearchPage.searchError" />
              </h2>
            ) : null}
            <SearchResultsPanel
              className={css.searchListingsPanel}
              listings={listings}
              pagination={listingsAreLoaded ? pagination : null}
              search={searchParamsForPagination}
              setActiveListing={onActivateListing}
            />
          </div>
        )}
      </div>
    );
  }
}

MainPanel.defaultProps = {
  className: null,
  rootClassName: null,
  listings: [],
  resultsCount: 0,
  pagination: null,
  searchParamsForPagination: {},
  primaryFilters: null,
  secondaryFilters: null,
};

MainPanel.propTypes = {
  className: string,
  rootClassName: string,

  urlQueryParams: object.isRequired,
  listings: array,
  searchInProgress: bool.isRequired,
  searchListingsError: propTypes.error,
  searchParamsAreInSync: bool.isRequired,
  onActivateListing: func.isRequired,
  onManageDisableScrolling: func.isRequired,
  onOpenModal: func.isRequired,
  onCloseModal: func.isRequired,
  onMapIconClick: func.isRequired,
  pagination: propTypes.pagination,
  searchParamsForPagination: object,
  showAsModalMaxWidth: number.isRequired,
  customURLParamToConfig: object.isRequired,
  primaryFilters: object,
  secondaryFilters: object,
};

export default MainPanel;
