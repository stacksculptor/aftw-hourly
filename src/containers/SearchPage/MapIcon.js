import React, { PropTypes } from 'react';

const MapIcon = props => {
  const { className } = props;

  return (
    <svg
      className={className}
      width="27"
      height="27"
      viewBox="0 0 27 27"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <filter
          x="-3.8%"
          y="-10.4%"
          width="107.7%"
          height="129.2%"
          filterUnits="objectBoundingBox"
          id="a"
        >
          <feOffset dy="2" in="SourceAlpha" result="shadowOffsetOuter1" />
          <feGaussianBlur stdDeviation="2" in="shadowOffsetOuter1" result="shadowBlurOuter1" />
          <feColorMatrix
            values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.2 0"
            in="shadowBlurOuter1"
            result="shadowMatrixOuter1"
          />
          <feMerge>
            <feMergeNode in="shadowMatrixOuter1" /><feMergeNode in="SourceGraphic" />
          </feMerge>
        </filter>
      </defs>
      <g
        filter="url(#a)"
        transform="translate(-41 -12)"
        strokeWidth="1.4"
        fill="none"
        fillRule="evenodd"
        strokeLinecap="round"
        strokeLinejoin="round"
      >
        <path
          d="M51.667 32L46 27.952V15l5.667 4.048L57.333 15 63 19.048V32l-5.667-4.048zM51.7 19v13M57.3 28V15"
        />
      </g>
    </svg>
  );
};

const { string } = PropTypes;

MapIcon.defaultProps = {
  className: null,
};

MapIcon.propTypes = {
  className: string,
};

export default MapIcon;
