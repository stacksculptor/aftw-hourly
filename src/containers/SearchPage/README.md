# SearchPage

SearchPage component has roughly 3 sections inside its layout:
Topbar, MainPanel (for results and filters), and Map

So, rough JSX presentation is something like:
```jsx
<Page>
  <TopbarContainer />
  <MainPanel />
  <SearchMap />
<Page>
```

Searches can be made by each of these components.

* Topbar contains location search (LocationAutocompleteInput)
* MainPanel has Filters that can fine-tune current location search
* SearchMap can create new location searches when the map's bounding box changes
  (i.e. moving, zooming, etc.)

## Topbar

In contrast to other pages, Topbar gets `currentSearchParams` among other props. This makes it
possible for Topbar to take current filters into account.

## MainPanel

MainPanel has two functions: showing searchResults and showing filters. There are two sets of
filters that can be passed to it: `primaryFilters`, `secondaryFilters`.

Primary filters are filters that are shown always on top of SearchResults as dropdown-selections.
We recommend that only 1 - 3 primary filters are passed in since they start to take too much space
on narrow screens.

Secondary filters create one more button to the space containing primary filters: *More filters*.
This more-filters button opens up a SearchFiltersPanel component that can be changed to show those
extra filters passed to it.
**Note:** Currently, it doesn't contain any filter components by default, even if you pass in some
filter data. Creating those filter components is part of the customization process.

On the mobile layout, all the filters are shown in separate mobile filters panel.

## SearchMap

SearchMap listens to 'idle' event and SearchPage function `onIndle` can create a new location search if
SearchMap's bounds have changed enough.

## Other things to consider

### URL Params vs marketplace-custom-config.js

Filter handling needs some mapping between configured extended data. For example,
*marketplace-custom-config.js* defines all the categories used by the app, but that data needs to be in
format `pub_category=selectedCategory` when we talk with the API. (We also use this format in URL.)

This mapping is done at the beginning of the file:
```
const CATEGORY_URL_PARAM = 'pub_category';
const AMENITIES_URL_PARAM = 'pub_amenities';

const customURLParamToConfig = {
  [CATEGORY_URL_PARAM]: 'categories',
  [AMENITIES_URL_PARAM]: 'amenities',
};
```

### SeachPage schema / SEO

Schema is created inside `createSearchResultSchema` in *SearchPage.helpers.js*. It needs listings
and address to make meaningful JSON-LD presentation for search engines.
