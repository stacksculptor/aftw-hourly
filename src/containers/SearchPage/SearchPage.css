@import '../../marketplace.css';

:root {
  --containerHeight: calc(100vh - var(--topbarHeightDesktop));
}

.topbar {
  @media (--viewportMedium) {
    position: fixed;
    top: 0;
    left: 0;
    /* We need to raise Topbar above .container */
    z-index: 1;
  }
}

.container {
  /* Layout */
  width: 100%;
  flex-grow: 1;
  display: flex;

  @media (--viewportMedium) {
    position: relative;
    padding-top: var(--topbarHeightDesktop);
    min-height: var(--containerHeight);
  }
}

.searchResultContainer {
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  padding: 18px 24px 96px 24px;

  @media (--viewportMedium) {
    /**
     * .container is using flexbox,
     * This specifies that searchResultContainer is taking 60% from the viewport width
     */
    flex-basis: 50%;
  }

  @media (--viewportLarge) {
    flex-basis: 62.5%;
    padding: 24px 36px 36px 36px;
  }
}

.error {
  color: var(--failColor);
  padding-left: 24px;

  @media (--viewportMedium) {
    padding-left: 36px;
  }
}

.searchString {
  /* Search string should not break on white spaces - i.e. line-break should happen before. */
  white-space: nowrap;
}

.searchAttributes {
  margin-bottom: 16px;

  @media (--viewportLarge) {
    margin-bottom: 24px;
  }
}

.listings {
  flex-grow: 1;
  display: flex;
  flex-direction: column;
}

.newSearchInProgress {
  opacity: 0.1;
  transition: opacity var(--transitionStyleButton);
}

.searchListingsPanel {
  flex-grow: 1;
}

.mapPanel {
  @media (--viewportMedium) {
    /**
     * .container is using flexbox,
     * This specifies that mapPanel is taking 34% from the viewport width
     */
    flex-basis: 50%;

    /* Own layout settings */
    display: flex;
    height: 100%;
  }

  @media (--viewportLarge) {
    flex-basis: 37.5%;
  }
}

.map {
  width: 100vw;
  height: 100vh;

  @media (--viewportMedium) {
    /* Map is fixed so that it doesn't scroll along search results */
    position: fixed;
    top: var(--topbarHeightDesktop);
    right: 0;

    /* Fixed content needs width relative to viewport */
    width: 50vw;
    height: var(--containerHeight);
  }

  @media (--viewportLarge) {
    width: 37.5vw;
  }
}
