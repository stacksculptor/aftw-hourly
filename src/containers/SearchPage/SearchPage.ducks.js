/**
 * This file contains Action constants, Action creators, and reducer of a page
 * container. We are following Ducks module proposition:
 * https://github.com/erikras/ducks-modular-redux
 */
import { unionWith, isEqual } from 'lodash';
import { call, take, put } from 'redux-saga/effects';
import { createRequestTypes } from '../../util/sagaHelpers';

// ================ Action types ================ //

export const ADD_FILTER = 'app/SearchPage/ADD_FILTER';

// async actions use request - success / failure pattern
export const LOAD_LISTINGS = createRequestTypes('app/SearchPage/LOAD_LISTINGS');

// ================ Reducer ================ //

export const initialState = {
  filters: [],
  initialListingsLoaded: false,
  listings: [],
  loadingListings: false,
};

export default function reducer(state = initialState, action = {}) {
  const { type, payload } = action;

  switch (type) {
    case ADD_FILTER: {
      const stateFilters = state.filters || [];
      return { ...state, ...{ filters: unionWith(stateFilters, [payload], isEqual) } };
    }
    case LOAD_LISTINGS.REQUEST:
    case LOAD_LISTINGS.SUCCESS:
    case LOAD_LISTINGS.FAILURE:
      return { ...state, ...payload };
    default:
      return state;
  }
}

// ================ Action creators ================ //

export const addFilter = (key, value) => ({ type: ADD_FILTER, payload: { [key]: value } });

// async actions use request - success - failure pattern
const action = (type, payload = {}) => ({ type, ...payload });

export const loadListings = {
  request: () =>
    action(LOAD_LISTINGS.REQUEST, {
      payload: { loadingListings: true, initialListingsLoaded: false },
    }),
  success: listings =>
    action(LOAD_LISTINGS.SUCCESS, {
      payload: { loadingListings: false, initialListingsLoaded: true, listings },
    }),
  failure: error =>
    action(LOAD_LISTINGS.FAILURE, {
      payload: { loadingListings: false, loadListingError: error, error: true },
    }),
};

// ================ Worker sagas ================ //

export function* callFetchListings(sdk) {
  const { response, error } = yield call(sdk.fetchListings);
  if (response) {
    yield put(loadListings.success(response));
  } else {
    yield put(loadListings.failure(error));
  }
}

// ================ Watcher sagas ================ //

export function* watchLoadListings(sdk) {
  // eslint-disable-next-line no-constant-condition
  while (true) {
    yield take(LOAD_LISTINGS.REQUEST);
    yield call(callFetchListings, sdk);
  }
}
