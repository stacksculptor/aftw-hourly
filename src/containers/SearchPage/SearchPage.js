import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { withRouter } from 'react-router-dom';
import debounce from 'lodash/debounce';
import isEqual from 'lodash/isEqual';
import unionWith from 'lodash/unionWith';
import classNames from 'classnames';
import config from '../../config';
import routeConfiguration from '../../routeConfiguration';
import {
  googleLatLngToSDKLatLng,
  googleBoundsToSDKBounds,
  sdkBoundsToFixedCoordinates,
  hasSameSDKBounds,
} from '../../util/googleMaps';
import { createResourceLocatorString } from '../../util/routes';
import { parse, stringify } from '../../util/urlHelpers';
import { propTypes } from '../../util/types';
import { getListingsById } from '../../ducks/marketplaceData.duck';
import { manageDisableScrolling, isScrollingDisabled } from '../../ducks/UI.duck';
import { SearchMap, ModalInMobile, Page } from '../../components';
import { TopbarContainer } from '../../containers';

import { searchListings, searchMapListings, setActiveListing } from './SearchPage.duck';
import {
  pickSearchParamsOnly,
  validURLParamsForExtendedData,
  validFilterParams,
  createSearchResultSchema,
} from './SearchPage.helpers';
import MainPanel from './MainPanel';
import css from './SearchPage.css';

// Pagination page size might need to be dynamic on responsive page layouts
// Current design has max 3 columns 12 is divisible by 2 and 3
// So, there's enough cards to fill all columns on full pagination pages
const RESULT_PAGE_SIZE = 24;
const MODAL_BREAKPOINT = 768; // Search is in modal on mobile layout
const SEARCH_WITH_MAP_DEBOUNCE = 300; // Little bit of debounce before search is initiated.
const BOUNDS_FIXED_PRECISION = 8;

export class SearchPageComponent extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isSearchMapOpenOnMobile: props.tab === 'map',
      isMobileModalOpen: false,
    };

    // Initiating map creates 'bounds_changes' event
    // we listen to that event to make new searches
    // So, if the search comes from location search input (this.viewportBounds == null),
    // we need to by pass extra searches created by Google Map's 'indle' event.
    // This is done by keeping track of map's viewport bounds (which differ from location bounds)
    this.viewportBounds = null;
    this.modalOpenedBoundsChange = false;
    this.searchMapListingsInProgress = false;

    this.filters = this.filters.bind(this);
    this.onIdle = debounce(this.onIdle.bind(this), SEARCH_WITH_MAP_DEBOUNCE);
    this.onOpenMobileModal = this.onOpenMobileModal.bind(this);
    this.onCloseMobileModal = this.onCloseMobileModal.bind(this);
  }

  filters() {
    const { categories, amenities } = this.props;

    return {
      categoryFilter: {
        paramName: 'pub_category',
        options: categories,
      },
      amenitiesFilter: {
        paramName: 'pub_amenities',
        options: amenities,
      },
    };
  }

  componentWillReceiveProps(nextProps) {
    if (!isEqual(this.props.location, nextProps.location)) {
      // If no mapSearch url parameter is given, this is original location search
      const { mapSearch } = parse(nextProps.location.search, {
        latlng: ['origin'],
        latlngBounds: ['bounds'],
      });
      if (!mapSearch) {
        this.viewportBounds = null;
      }
    }
  }

  // We are using Google Maps idle event instead of bounds_changed, since it will not be fired
  // too often (in the middle of map's pan or zoom activity)
  onIdle(googleMap) {
    const { history, location } = this.props;

    // parse query parameters, including a custom attribute named category
    const { address, bounds, mapSearch, ...rest } = parse(location.search, {
      latlng: ['origin'],
      latlngBounds: ['bounds'],
    });

    const viewportGMapBounds = googleMap.getBounds();
    const viewportBounds = sdkBoundsToFixedCoordinates(
      googleBoundsToSDKBounds(viewportGMapBounds),
      BOUNDS_FIXED_PRECISION
    );

    // ViewportBounds from (previous) rendering differ from viewportBounds currently set to map
    // I.e. user has changed the map somehow: moved, panned, zoomed, resized
    const viewportBoundsChanged =
      this.viewportBounds && !hasSameSDKBounds(this.viewportBounds, viewportBounds);

    // If mapSearch url param is given (and we have not just opened mobile map modal)
    // or original location search is rendered once,
    // we start to react to 'bounds_changed' event by generating new searches
    if (viewportBoundsChanged && !this.modalOpenedBoundsChange) {
      const originMaybe = config.sortSearchByDistance
        ? { origin: googleLatLngToSDKLatLng(viewportGMapBounds.getCenter()) }
        : {};
      const searchParams = {
        address,
        ...originMaybe,
        bounds: viewportBounds,
        mapSearch: true,
        ...validFilterParams(rest, this.filters()),
      };

      this.viewportBounds = viewportBounds;
      history.push(
        createResourceLocatorString('SearchPage', routeConfiguration(), {}, searchParams)
      );
    } else {
      this.viewportBounds = viewportBounds;
      this.modalOpenedBoundsChange = false;
    }
  }

  // Invoked when a modal is opened from a child component,
  // for example when a filter modal is opened in mobile view
  onOpenMobileModal() {
    this.setState({ isMobileModalOpen: true });
  }

  // Invoked when a modal is closed from a child component,
  // for example when a filter modal is opened in mobile view
  onCloseMobileModal() {
    this.setState({ isMobileModalOpen: false });
  }

  render() {
    const {
      intl,
      listings,
      location,
      mapListings,
      onManageDisableScrolling,
      pagination,
      scrollingDisabled,
      searchInProgress,
      searchListingsError,
      searchParams,
      activeListingId,
      onActivateListing,
    } = this.props;
    // eslint-disable-next-line no-unused-vars
    const { mapSearch, page, ...searchInURL } = parse(location.search, {
      latlng: ['origin'],
      latlngBounds: ['bounds'],
    });

    const filters = this.filters();

    // urlQueryParams doesn't contain page specific url params
    // like mapSearch, page or origin (origin depends on config.sortSearchByDistance)
    const urlQueryParams = pickSearchParamsOnly(searchInURL, filters);

    // Page transition might initially use values from previous search
    const urlQueryString = stringify(urlQueryParams);
    const paramsQueryString = stringify(pickSearchParamsOnly(searchParams, filters));
    const searchParamsAreInSync = urlQueryString === paramsQueryString;

    const validQueryParams = validURLParamsForExtendedData(searchInURL, filters);

    const isWindowDefined = typeof window !== 'undefined';
    const isMobileLayout = isWindowDefined && window.innerWidth < MODAL_BREAKPOINT;
    const shouldShowSearchMap =
      !isMobileLayout || (isMobileLayout && this.state.isSearchMapOpenOnMobile);

    const onMapIconClick = () => {
      this.useLocationSearchBounds = true;
      this.modalOpenedBoundsChange = true;
      this.setState({ isSearchMapOpenOnMobile: true });
    };

    const { address, bounds, origin } = searchInURL || {};
    const { title, description, schema } = createSearchResultSchema(listings, address, intl);

    // Set topbar class based on if a modal is open in
    // a child component
    const topbarClasses = this.state.isMobileModalOpen
      ? classNames(css.topbarBehindModal, css.topbar)
      : css.topbar;

    // N.B. openMobileMap button is sticky.
    // For some reason, stickyness doesn't work on Safari, if the element is <button>
    /* eslint-disable jsx-a11y/no-static-element-interactions */
    return (
      <Page
        scrollingDisabled={scrollingDisabled}
        description={description}
        title={title}
        schema={schema}
      >
        <TopbarContainer
          className={topbarClasses}
          currentPage="SearchPage"
          currentSearchParams={urlQueryParams}
        />
        <div className={css.container}>
          <MainPanel
            urlQueryParams={validQueryParams}
            listings={listings}
            searchInProgress={searchInProgress}
            searchListingsError={searchListingsError}
            searchParamsAreInSync={searchParamsAreInSync}
            onActivateListing={onActivateListing}
            onManageDisableScrolling={onManageDisableScrolling}
            onOpenModal={this.onOpenMobileModal}
            onCloseModal={this.onCloseMobileModal}
            onMapIconClick={onMapIconClick}
            pagination={pagination}
            searchParamsForPagination={parse(location.search)}
            showAsModalMaxWidth={MODAL_BREAKPOINT}
            primaryFilters={{
              categoryFilter: filters.categoryFilter,
              amenitiesFilter: filters.amenitiesFilter,
            }}
          />
          <ModalInMobile
            className={css.mapPanel}
            id="SearchPage.map"
            isModalOpenOnMobile={this.state.isSearchMapOpenOnMobile}
            onClose={() => this.setState({ isSearchMapOpenOnMobile: false })}
            showAsModalMaxWidth={MODAL_BREAKPOINT}
            onManageDisableScrolling={onManageDisableScrolling}
          >
            <div className={css.mapWrapper}>
              {shouldShowSearchMap ? (
                <SearchMap
                  reusableContainerClassName={css.map}
                  activeListingId={activeListingId}
                  bounds={bounds}
                  center={origin}
                  listings={mapListings || []}
                  onIdle={this.onIdle}
                  isOpenOnModal={this.state.isSearchMapOpenOnMobile}
                  onCloseAsModal={() => {
                    onManageDisableScrolling('SearchPage.map', false);
                  }}
                  useLocationSearchBounds={!this.viewportBounds}
                />
              ) : null}
            </div>
          </ModalInMobile>
        </div>
      </Page>
    );
    /* eslint-enable jsx-a11y/no-static-element-interactions */
  }
}

SearchPageComponent.defaultProps = {
  listings: [],
  mapListings: [],
  pagination: null,
  searchListingsError: null,
  searchParams: {},
  tab: 'listings',
  categories: config.custom.categories,
  amenities: config.custom.amenities,
  activeListingId: null,
};

const { array, bool, func, oneOf, object, shape, string } = PropTypes;

SearchPageComponent.propTypes = {
  listings: array,
  mapListings: array,
  onActivateListing: func.isRequired,
  onManageDisableScrolling: func.isRequired,
  onSearchMapListings: func.isRequired,
  pagination: propTypes.pagination,
  scrollingDisabled: bool.isRequired,
  searchInProgress: bool.isRequired,
  searchListingsError: propTypes.error,
  searchParams: object,
  tab: oneOf(['filters', 'listings', 'map']).isRequired,
  categories: array,
  amenities: array,

  // from withRouter
  history: shape({
    push: func.isRequired,
  }).isRequired,
  location: shape({
    search: string.isRequired,
  }).isRequired,

  // from injectIntl
  intl: intlShape.isRequired,
};

const mapStateToProps = state => {
  const {
    currentPageResultIds,
    pagination,
    searchInProgress,
    searchListingsError,
    searchParams,
    searchMapListingIds,
    activeListingId,
  } = state.SearchPage;
  const pageListings = getListingsById(state, currentPageResultIds);
  const mapListings = getListingsById(
    state,
    unionWith(currentPageResultIds, searchMapListingIds, (id1, id2) => id1.uuid === id2.uuid)
  );

  return {
    listings: pageListings,
    mapListings,
    pagination,
    scrollingDisabled: isScrollingDisabled(state),
    searchInProgress,
    searchListingsError,
    searchParams,
    activeListingId,
  };
};

const mapDispatchToProps = dispatch => ({
  onManageDisableScrolling: (componentId, disableScrolling) =>
    dispatch(manageDisableScrolling(componentId, disableScrolling)),
  onSearchMapListings: searchParams => dispatch(searchMapListings(searchParams)),
  onActivateListing: listingId => dispatch(setActiveListing(listingId)),
});

// Note: it is important that the withRouter HOC is **outside** the
// connect HOC, otherwise React Router won't rerender any Route
// components since connect implements a shouldComponentUpdate
// lifecycle hook.
//
// See: https://github.com/ReactTraining/react-router/issues/4671
const SearchPage = compose(withRouter, connect(mapStateToProps, mapDispatchToProps), injectIntl)(
  SearchPageComponent
);

SearchPage.loadData = (params, search) => {
  const queryParams = parse(search, {
    latlng: ['origin'],
    latlngBounds: ['bounds'],
  });
  const { page = 1, address, origin, ...rest } = queryParams;
  const originMaybe = config.sortSearchByDistance && origin ? { origin } : {};
  return searchListings({
    ...rest,
    ...originMaybe,
    page,
    perPage: RESULT_PAGE_SIZE,
    include: ['author', 'images'],
    'fields.image': ['variants.landscape-crop', 'variants.landscape-crop2x'],
    'limit.images': 1,
  });
};

export default SearchPage;
