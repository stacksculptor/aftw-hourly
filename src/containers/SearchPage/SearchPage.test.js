import React from 'react';
import { renderShallow } from '../../util/test-helpers';
import { SearchPageComponent } from './SearchPage';
import reducer, { ADD_FILTER, addFilter, initialState } from './SearchPage.ducks';

describe('SearchPageComponent', () => {
  it('matches snapshot', () => {
    const tree = renderShallow(<SearchPageComponent onLoadListings={v => v} />);
    expect(tree).toMatchSnapshot();
  });
});

describe('SearchPageDucs', () => {
  describe('actions', () => {
    it('should create an action to add a filter', () => {
      const expectedAction = { type: ADD_FILTER, payload: { location: 'helsinki' } };
      expect(addFilter('location', 'helsinki')).toEqual(expectedAction);
    });
  });

  describe('reducer', () => {
    it('should return the initial state', () => {
      const initial = reducer(undefined, {});
      expect(initial).toEqual(initialState);
    });

    it('should handle ADD_FILTER', () => {
      const addFilter1 = addFilter('location', 'helsinki');
      const addFilter2 = addFilter('gears', 3);
      const reduced = reducer([], addFilter1);
      const reducedWithInitialContent = reducer({ filters: [addFilter1.payload] }, addFilter2);
      expect(reduced).toEqual({ filters: [addFilter1.payload] });
      expect(reducedWithInitialContent).toEqual({
        filters: [addFilter1.payload, addFilter2.payload],
      });
    });

    it('should handle duplicates ADD_FILTER', () => {
      const filter = { location: 'helsinki' };
      const addFilter = { type: ADD_FILTER, payload: filter };
      const reducedWithInitialContent = reducer({ filters: [filter] }, addFilter);
      expect(reducedWithInitialContent).toEqual({ filters: [filter] });
    });
  });
});
