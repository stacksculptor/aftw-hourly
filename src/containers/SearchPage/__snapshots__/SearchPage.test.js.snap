// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPageComponent matches snapshot 1`] = `
<Page
  description="SearchPage.schemaDescription"
  schema={
    Object {
      "@context": "http://schema.org",
      "@type": "SearchResultsPage",
      "description": "SearchPage.schemaDescription",
      "mainEntity": Array [
        "{\\"@type\\":\\"ItemList\\",\\"name\\":\\"SearchPage.schemaMapSearch\\",\\"itemListOrder\\":\\"http://schema.org/ItemListOrderAscending\\",\\"itemListElement\\":[]}",
      ],
      "name": "SearchPage.schemaTitle",
    }
  }
  scrollingDisabled={false}
  title="SearchPage.schemaTitle"
>
  <withRouter(Connect(TopbarContainerComponent)) />
  <div>
    <div>
      <withRouter(InjectIntl(SearchFiltersComponent))
        categories={
          Array [
            Object {
              "key": "cat1",
              "label": "Cat 1",
            },
            Object {
              "key": "cat2",
              "label": "Cat 2",
            },
          ]
        }
        features={
          Array [
            Object {
              "key": "towels",
              "label": "Towels",
            },
            Object {
              "key": "bathroom",
              "label": "Bathroom",
            },
            Object {
              "key": "swimming_pool",
              "label": "Swimming pool",
            },
            Object {
              "key": "own_drinks",
              "label": "Own drinks allowed",
            },
            Object {
              "key": "jacuzzi",
              "label": "Jacuzzi",
            },
            Object {
              "key": "audiovisual_entertainment",
              "label": "Audiovisual entertainment",
            },
            Object {
              "key": "barbeque",
              "label": "Barbeque",
            },
            Object {
              "key": "own_food_allowed",
              "label": "Own food allowed",
            },
          ]
        }
        listingsAreLoaded={true}
        onManageDisableScrolling={[Function]}
        onMapIconClick={[Function]}
        resultsCount={22}
        searchInProgress={false}
        searchListingsError={null}
        urlQueryParams={
          Object {
            "address": undefined,
            "bounds": undefined,
            "origin": undefined,
            "pub_category": undefined,
            "pub_features": undefined,
          }
        }
      />
      <InjectIntl(withRouter(SearchFiltersMobileComponent))
        categories={
          Array [
            Object {
              "key": "cat1",
              "label": "Cat 1",
            },
            Object {
              "key": "cat2",
              "label": "Cat 2",
            },
          ]
        }
        listingsAreLoaded={true}
        onCloseModal={[Function]}
        onManageDisableScrolling={[Function]}
        onMapIconClick={[Function]}
        onOpenModal={[Function]}
        resultsCount={22}
        searchInProgress={false}
        searchListingsError={null}
        showAsModalMaxWidth={768}
        urlQueryParams={
          Object {
            "address": undefined,
            "bounds": undefined,
            "origin": undefined,
            "pub_category": undefined,
            "pub_features": undefined,
          }
        }
      />
      <div
        className=""
      >
        <SearchResultsPanel
          className={null}
          listings={Array []}
          pagination={
            Object {
              "page": 1,
              "perPage": 12,
              "totalItems": 22,
              "totalPages": 2,
            }
          }
          rootClassName={null}
          search={Object {}}
        />
      </div>
    </div>
    <withViewport(ModalInMobileComponent)
      id="SearchPage.map"
      isModalOpenOnMobile={false}
      onClose={[Function]}
      onManageDisableScrolling={[Function]}
      showAsModalMaxWidth={768}
    >
      <div>
        <SearchMapComponent
          bounds={null}
          center={
            LatLng {
              "lat": 0,
              "lng": 0,
            }
          }
          className={null}
          coordinatesConfig={
            Object {
              "circleOptions": Object {
                "fillColor": "#c0392b",
                "fillOpacity": 0.2,
                "strokeColor": "#c0392b",
                "strokeWeight": 0.5,
              },
              "circleRadius": 500,
              "fuzzy": false,
              "fuzzyDefaultZoomLevel": 14,
            }
          }
          isOpenOnModal={false}
          listings={Array []}
          mapRootClassName={null}
          onCloseAsModal={[Function]}
          onIdle={[Function]}
          rootClassName={null}
          useLocationSearchBounds={true}
          zoom={11}
        />
      </div>
    </withViewport(ModalInMobileComponent)>
  </div>
</Page>
`;
